File: src/main/java/com/arcao/geocaching/api/oauth/services/ServerTimestampServiceImpl.java
package com.arcao.geocaching.api.oauth.services;

import com.arcao.wherigoservice.api.WherigoApiFactory;
import java.util.Random;
import org.scribe.services.TimestampServiceImpl;
import timber.log.Timber;

public class ServerTimestampServiceImpl extends TimestampServiceImpl {
	private final Random rand = new Random();
	private final long ts;

	public ServerTimestampServiceImpl() {
		long time;
		try {
			time = WherigoApiFactory.create().getTime();
			Timber.i("server time received (ms): " + time);
		} catch (Exception e) {
			time = System.currentTimeMillis();
			Timber.e(e, "No server time received. Used system time (ms): " + time);
		}

		// timestamp in seconds
		ts = time / 1000;
	}

	@Override
	public String getTimestampInSeconds() {
		return String.valueOf(ts);
	}

	@Override
	public String getNonce() {
		return String.valueOf(ts + rand.nextInt());
	}
}

File: app/src/main/java/com/arcao/geocaching/api/oauth/services/ServerTimestampServiceImpl.kt
package com.arcao.geocaching.api.oauth.services

import com.arcao.wherigoservice.api.WherigoService
import com.github.scribejava.core.services.TimestampServiceImpl
import org.koin.core.KoinComponent
import org.koin.core.inject
import timber.log.Timber
import java.util.Random

class ServerTimestampServiceImpl : TimestampServiceImpl(), KoinComponent {
    private val wherigoService by inject<WherigoService>()

    private val rand = Random()

    private val ts: Long by lazy {
        val time = try {
            wherigoService.getTime().also {
                Timber.i("server time received (ms): %d", it)
            }
        } catch (e: Exception) {
            System.currentTimeMillis().also {
                Timber.e(e, "No server time received. Used system time (ms): %d", it)
            }
        }

        // timestamp in seconds
        time / 1000
    }

    override fun getTimestampInSeconds(): String {
        return ts.toString()
    }

    override fun getNonce(): String {
        return (ts + rand.nextInt()).toString()
    }
}
